brave/brave-ios
# Fix #3227: Implement History Sync

## Summary of Changes

This Pull Request contains 

- History Migration
- Sync Profile
- History Fetch/Add/Delete with Core and Sync
- Brave-Core Bump 1.26.x
- Delete All functionality on History
- Empty State for History 

###  History Migration:

We are adding history brave-core migration to our existing migrator used for bookmarks and making both of the sync types work together and handle error cases related for both of the types. So in future the new sync types can also use same migrator.

### History Fetch/Add/Delete/Remove with Core and Sync

A new HistoryFetcher is created where we perform the fetch from brave-core and various utility methods are created over the data. This fetcher is using Ordered Dictionary so in order to use this we added [Swift Collections](https://github.com/apple/swift-collections) package. Believe the utility brought with Swift Collections will also be useful for the rest of the application. https://swift.org/blog/swift-collections/

In addition old add delete methods are replaced with brave-core  functionality. 

### Sync Profile

Enable/Disable different Sync Types is added. Sync Settings will contain a new section which will allow user to toggle different Sync types according to users preference. This is using BraveSyncProfileService and it will also help when new sync types are added to give user ability to turn them on/off.


This pull request fixes #3227
This pull request fixes #3723
This pull request fixes #3721
This pull request fixes #3869
This pull request fixes #3591

Brave Core PR is merged and uplifted to 1.26.x and in addition this PR also handles the changes made in core to align history and bookmarks code.

Core PR For History Sync: https://github.com/brave/brave-core/pull/8869

Core PR For alignment: https://github.com/brave/brave-core/pull/9145

SecReview: https://github.com/brave/security/issues/501

## Submitter Checklist:


## Test Plan:

Test Migration: The test cases and conditions are same with bookmark. Try migration 3 times but for both history and bookmarks together in this case.

- Update application with existing History base
- Check migration is happened and data is same

Normal Add / Delete / Delete All 

- Fresh install or updated app
- Try Add history by entering urls or redirecting from other places
- Delete individual history
- Delete All history in History Screen

Test Sync: Same testing cases with bookmarks release but this time after joining to sync chain can actually turn on/off 
which type of sync should be enabled

- Can test turning on different sync types and check If their functionality can be turned on/off
- History Sync should be tested but keep in mind only typed urls is going to be synced from desktop to mobile also mobile to desktop 
- Migrated urls will be marked as types since we dont have informationa bout them so they should be synced

## Screenshots:

![119523015-6b475a80-bd4a-11eb-99f3-f2daf41bf66a](https://user-images.githubusercontent.com/6643505/120040837-3c90e480-bfd5-11eb-9e3e-1c47e94d008b.png)

![33 3333233](https://user-images.githubusercontent.com/6643505/120041067-a27d6c00-bfd5-11eb-847e-ba4902c14ba1.png)


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
